Projects/Project12B UserDefaults Codable/Project10 Names to Faces/ViewController.swift

```diff
@@ -7,46 +7,127 @@
         super.viewDidLoad()
-        load()
-        setupNavigationItems()
+        
+        title = "Locked"
+        
+        NotificationCenter.default.addObserver(self, selector: #selector(lockApp), name: UIApplication.willResignActiveNotification, object: nil)
+        
+        authenticateButton = UIBarButtonItem(title: "Authenticate", style: .plain, target: self, action: #selector(authenticateTapped))
+        addFromPhotosButton = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addNewPerson))
+        addFromCameraButton = UIBarButtonItem(barButtonSystemItem: .camera, target: self, action: #selector(addNewPersonByCamera))
+        
+        authenticateButton?.isEnabled = true
+        addFromCameraButton?.isEnabled = false
+        addFromPhotosButton?.isEnabled = false
+        
+        if UIImagePickerController.isSourceTypeAvailable(.camera) {
+            navigationItem.rightBarButtonItems = [addFromPhotosButton!, addFromCameraButton!]
+        } else {
+            navigationItem.rightBarButtonItem = addFromPhotosButton
+        }
+        
+        navigationItem.leftBarButtonItem = authenticateButton
     }
     
-    @objc func addNewPerson() {
-        let picker = UIImagePickerController()
-        picker.allowsEditing = true
-        picker.delegate = self
-        present(picker, animated: true)
+    @objc func lockApp() {
+        title = "Locked"
+        authenticateButton?.isEnabled = true
+        addFromCameraButton?.isEnabled = false
+        addFromPhotosButton?.isEnabled = false
+        people = []
+        self.collectionView.reloadData()
     }
     
-    @objc func addNewPersonByCamera() {
-        let picker = UIImagePickerController()
-        picker.sourceType = .camera
-        picker.allowsEditing = true
-        picker.delegate = self
-        present(picker, animated: true)
+    @objc func authenticateTapped() {
+        let context = LAContext()
+        var error: NSError?
+        if context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) {
+            let reason = "Identify yourself with touchID"
+            context.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: reason) { [weak self] (success, error) in
+                if success {
+                    // authentication might not happen on the main thread so use GCD
+                    DispatchQueue.main.async {
+                        self?.loadPeople()
+                        self?.authenticateButton?.isEnabled = false
+                        self?.addFromCameraButton?.isEnabled = true
+                        self?.addFromPhotosButton?.isEnabled = true
+                        self?.title = "Unlocked"
+                    }
+                } else {
+                    // alert failed to authenticate using biometrics
+                    let ac = UIAlertController(title: "FaceID Failed", message: "Biometric faceID failed in recognizing your face.", preferredStyle: .alert)
```

Comment:
Have you checked what the error description is in this case, you might be able to use it rather than using a generic one.

Reply:
Ahhh okay perfect! I wasn't taking advantage of the error parameter passed into the closure...

context.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: reason) { [weak self] (success, error) in

}

I'm changing the generic message to use the error.localizedDescription to handle the error message displayed. For example, if the user presses cancel and after attempting a face identification the alert's message will now use the error "Canceled by user."